Move socket into separate directory

WhyNotHugo opened the pull request:
This is mostly to ease setup and configuration with sandboxed browsers.

The socket currently existing in `$XDG_RUNTIME_DIR`. When sandboxing a
browser, it would be unsafe to mount this directory inside the sandbox.
Mounting the socket into the sandbox's filesystem is also not possible
in cases where KeePassXC is [re]started after the browser has started.

This commit moves the socket into its own isolated subdirectory, which
can be safely mounted into sandboxes. Sandbox engines can create the
directory themselves (in case the browser starts before KeePassXC). Both
Flatpak and Firejail support this configuration.

A symlink is also created, linking the previous location to the new
location. This is meant for backwards compatibility and should
eventually be dropped.

The directory can't be named `org.keepassxc.KeePassXC.BrowserServer`,
since that would collide with the symlink. Instead, the directory has
been created to match the format used for Flatpak builds, which make it
a bit less of a snowflake build, while following accepted conventions.

Fixes: https://github.com/keepassxreboot/keepassxc/issues/8018

## Testing strategy

### Backwards compatibility

Try using browser integration on an existing, working, setup. It should all work without any issues.

### For sandboxed applications

See https://github.com/keepassxreboot/keepassxc/discussions/6741#discussioncomment-1154242, without tampering with `keepassxc.local`.

## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)
- ✅ New feature (change that adds functionality)

## Notes

It might make sense to drop this bit, since now it matches the default behaviour: https://github.com/keepassxreboot/keepassxc/blob/736517f4553b443a74a420ef421b7c62695e6462/src/browser/BrowserShared.cpp#L35-L37

Not sure if it's fine to do this on the PR of it it's best to do that separately.

Commit:
Move socket into separate directory

This is mostly to ease setup and configuration with sandboxed browsers.

The socket currently existing in `$XDG_RUNTIME_DIR`. When sandboxing a
browser, it would be unsafe to mount this directory inside the sandbox.
Mounting the socket into the sandbox's filesystem is also not possible
in cases where KeePassXC is [re]started after the browser has started.

This commit moves the socket into its own isolated subdirectory, which
can be safely mounted into sandboxes. Sandbox engines can create the
directory themselves (in case the browser starts before KeePassXC). Both
Flatpak and Firejail support this configuration.

A symlink is also created, linking the previous location to the new
location. This is meant for backwards compatibility and should
eventually be dropped.

The directory can't be named `org.keepassxc.KeePassXC.BrowserServer`,
since that would collide with the symlink. Instead, the directory has
been created to match the format used for Flatpak builds, which make it
a bit less of a snowflake build, while following accepted conventions.

Given that the preferred path now matches what Flatpak uses, the block
handling Flatpak and non-Flatpak is now the same.

If `$XDG_RUNTIME_DIR` is undefined, the temporary directory is used,
though reading the socket from this location is discouraged.

Closes: https://github.com/keepassxreboot/keepassxc/issues/8018
References: https://github.com/keepassxreboot/keepassxc/discussions/6741

droidmonkey commented:
Nice!

WhyNotHugo commented:
Squashed follow-up commits. Formatting issues are all fixed. Ready for review/feedback.

droidmonkey commented:
Interesting, our code format check seems to be broken @phoerious. We shouldn't be passing when there are if statements without { } blocks.

droidmonkey commented on src/browser/BrowserShared.cpp:
```diff
@@ -35,10 +36,25 @@ namespace BrowserShared
+        // Use XDG_RUNTIME_DIR if available, else use /tmp.
+        QString xdgRuntimePath = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
+        if (xdgRuntimePath.isEmpty())
+            return QStandardPaths::writableLocation(QStandardPaths::TempLocation) + serverName;
```
Besides wrapping this in { }, we should prefer the dedicated directory within the temp dir as well. Actually this would probably be better written as:
```
if (xdgRuntimePath.isEmpty()) {
    xdgRuntimePath = QStandardPaths::writableLocation(QStandardPaths::TempLocation);
}
```
Then let the rest of the function proceed as normal.

WhyNotHugo replied:
Done in afe8f5bed971469bc609428817453226e8bf3d9e.

droidmonkey commented on src/browser/BrowserShared.cpp:
```diff
@@ -35,10 +36,25 @@ namespace BrowserShared
+
+        // Put the socket in a dedicated directory.
+        // This directory will be easily mountable by sandbox containers.
+        QDir socketDir = xdgRuntimeDir.filePath("app/org.keepassxc.KeePassXC");
```
is the "app" directory a standard? Why not just use the fully qualified reverse domain?

WhyNotHugo replied:
I noticed that Flatpak used it, and it seemed sensible to align with it. But I don't think it's actually a "standard" thing outside of Flatpak. I can drop `app/` if you wan't; I'm fine either way.

droidmonkey replied:
good point, in fact we can get rid of the flatpak special case with this change, perhaps wrap the symlink part with a #ifndef KEEPASSXC_DIST_FLATPAK

WhyNotHugo replied:
Also done. Not sure if the formatting is correct for nested `#elif`... seems a bit ambiguous. Lemme know if it needs to be different.

droidmonkey commented on src/browser/BrowserShared.cpp:
```diff
@@ -35,10 +36,25 @@ namespace BrowserShared
+        if (xdgRuntimePath.isEmpty())
+            return QStandardPaths::writableLocation(QStandardPaths::TempLocation) + serverName;
+
+        QDir xdgRuntimeDir = QDir(xdgRuntimePath);
```
You don't need this variable. Just create the socketDir variable directly using `QDir(xdgRuntimePath + "/app/org.keepassxc.KeePassXC")`

WhyNotHugo replied:
Done.

droidmonkey commented on src/browser/BrowserShared.cpp:
```diff
@@ -35,10 +36,25 @@ namespace BrowserShared
+        // This directory will be easily mountable by sandbox containers.
+        QDir socketDir = xdgRuntimeDir.filePath("app/org.keepassxc.KeePassXC");
+        if (!socketDir.exists())
+            xdgRuntimeDir.mkpath("app/org.keepassxc.KeePassXC");
```
same, use `socketDir.mkPath({})`. No need to check for existence, the function does that for you and returns true if it exists.

WhyNotHugo commented:
Looking at the [logs for the build](https://ci.keepassxc.org/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=212571&_focus=1247), I noticed this:

> [13:52:59][Step 3/4] QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-keepassxc'

It turns out that [if `XDG_RUNTIME_DIR` is unset, then `QStandardsPaths::RuntimeLocation` returns fallback a path in `/tmp`](https://github.com/qt/qtbase/blob/ce0d7477a91d73a642bf1f3b34eb7e89f4dfab9e/src/corelib/io/qstandardpaths_unix.cpp#L256-L277). This implies that it only returns empty if a path inside `/tmp` was unusable, so the fallback code that's already present in `src/browser/BrowserShared.cpp` is wrong, since it'll try to use `/tmp` in cases when `/tmp` is unusable.

The fallback should be dropped. It's clear that nobody is relying on this fallback for the extension tho, because, unless I'm missing something, it cannot work.

Edit: implemented in https://github.com/keepassxreboot/keepassxc/pull/8030/commits/a455ee49bd4663d0f041a25dcd676fe0479f9b2b

---

That aside, I've no idea why the Ubuntu build is failing. Is this a fluke? I'll try to re-trigger CI and see if it passes.

Edit: it passed.

Commit:
Remove bogus fallback

It if XDG_RUNTIME_DIR is unset, then QStandardsPaths::RuntimeLocation
returns fallback a path in /tmp. This implies that the fallback code
remove here is wrong, since the fallback to /tmp is already handled by
Qt and doesn't need to be done again on the KeePassXC side.

It's clear that nobody is relying on the existing fallback for the
extension, since the code-path that creates triggers this fallback was
unreachable.

droidmonkey commented:
Agree the fallback is silly. Perhaps we can issue a warning to the console if the socket cannot be created.

WhyNotHugo commented:
> Agree the fallback is silly. Perhaps we can issue a warning to the console if the socket cannot be created.

Can we address it separately? I'm not really sure how to do that, and I think this PR is ready as-is.

droidmonkey commented:
Yah no worries

droidmonkey commented:
I'll be looking at prs later this week

droidmonkey commented on src/browser/BrowserShared.cpp:
```diff
@@ -31,14 +32,22 @@ namespace BrowserShared
 #elif defined(Q_OS_UNIX) && !defined(Q_OS_MACOS)
-        // Use XDG_RUNTIME_DIR instead of /tmp if it's available
+        // This returns XDG_RUNTIME_DIR or else a temporary subdirectory.
         QString path = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
```
Need to append a '/' to this variable otherwise the symlink will fail.

WhyNotHugo replied:
When creating the symlink `path` is concatenated to `serverName`, and `serverName` has a leading slash.

droidmonkey replied:
It sure does! I hang my head in the deepest shame

droidmonkey commented on src/browser/BrowserShared.cpp:
```diff
@@ -31,14 +32,22 @@ namespace BrowserShared
+
+        // Put the socket in a dedicated directory.
+        // This directory will be easily mountable by sandbox containers.
+        QString subPath = path + "/app/org.keepassxc.KeePassXC/";
```
Once you append the trailing slash above, you can drop it from the start of this string.

WhyNotHugo replied:
Given that `serverName` has a leading slash, I followed suit here and did the same. Otherwise all the concatenations get muddier.

WhyNotHugo commented:
Thanks!

WhyNotHugo commented:
Hmmm, the directory is created with permission 755, but keepass refuses to create the socket due to directory permissions (requires them to be 700).

I'm not that familiar with Qt, can we change `mkdir` so that is uses 0700, or should I just add a `chmod`?

droidmonkey commented:
Use https://doc.qt.io/qt-5/qfileinfo.html#permission

WhyNotHugo commented:
Oh, never mind. I was running in a sandbox and the permissions of `$XDG_RUNTIME_DIR` were wrong. Nothing to be fixed here, my bad.
